### PR TITLE
[netcore] Implement AppDomain.TypeResolve

### PR DESCRIPTION
**Pull request by CoffeeFlux**
In sre.c, avoid a crash due to the missing event but don't properly fix things up (since it's broken anyways). Otherwise, straightforward.

### CODE REVIEW

File: mono/metadata/sre.c [hunk 1]
```diff
@@ -4272,6 +4273,7 @@ ensure_complete_type (MonoClass *klass, MonoError *error)
 		// Asserting here could break a lot of code
 		//g_assert (klass->wastypebuilder);
 	}
+#endif
```
Comment (lambdageek) [inline mono/metadata/sre.c]: i'm a bit worried we'll never return to fix up this function once we're working on SRE.  maybe move the `#ifdef` under the `if` and add an `#else` case:

```
if (image_is_dynamic (...) && ...) {
#ifndef ENABLE_NETCORE
  ...
#else
  g_assert_not_reached ();
#endif
}
```

File: mono/metadata/appdomain.c [hunk 2]
```diff
@@ -855,20 +855,65 @@ mono_domain_try_type_resolve (MonoDomain *domain, char *name, MonoObject *typebu
 
 	MonoReflectionAssemblyHandle ret = NULL_HANDLE_INIT;
 
+	// This will not work correctly on netcore, both due to ALCs and the lack of a reflection-only context
```
[the remark this answers is not in the document]
Comment (CoffeeFlux) [reply mono/metadata/appdomain.c]: I believe the TypeBuilder event is a reflection-only thing? On netcore, it technically exists but nothing uses it.

File: mono/metadata/sre.c (hunk 1 above)
Comment (CoffeeFlux) [reply mono/metadata/sre.c]: Hmm, fair. Will do.

File: mono/metadata/appdomain.c [hunk 3]
```diff
@@ -811,8 +811,12 @@ mono_domain_create_appdomain_internal (char *friendly_name, MonoAppDomainSetupHa
 gboolean
 mono_domain_has_type_resolve (MonoDomain *domain)
 {
+	/*pedump doesn't create an appdomain, so the domain object doesn't exist.*/
+	if (!domain->domain)
```
[the remark this answers is not in the document]
Comment (CoffeeFlux) [reply mono/metadata/appdomain.c]: Will update.

File: mono/metadata/appdomain.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (lambdageek) [reply mono/metadata/appdomain.c]: Ah you meant "reflection only" in the sense of "used only by reflection APIs" not "is only used by the reflection-only context".

`TypeBuilder` is part of System.Reflection.Emit.  On netcore, SRE works a bit differently that on .NET Framework - you can use S.R.E.TypeBuilder to create new types at runtime, but you cannot save them to disk. So in Mono terms: we will need to get `sre.c` working as part of the netcore work, but we won't need `sre-save.c`

[the remark this answers is not in the document]
Comment (CoffeeFlux) [reply mono/metadata/appdomain.c]: I was confused on what this event was for. Updated.

File: mono/metadata/appdomain.c [hunk 4]
```diff
@@ -855,20 +855,70 @@ mono_domain_try_type_resolve (MonoDomain *domain, char *name, MonoObject *typebu
...
+
+	goto_if_nok (error, return_null);
+
+	g_assert (domain);
+	g_assert (MONO_HANDLE_BOOL (name));
+
+	if (mono_runtime_get_no_exec ())
+		goto return_null;
+
+	MonoReflectionAssemblyHandle assembly_handle;
```
Comment (jaykrell) [inline mono/metadata/appdomain.c]: Goto around init.

**Pull request merged**